jaxlib/lapack.pyx

```diff
@@ -77,7 +78,11 @@ cdef void blas_strsm(void* out, void** data) nogil:
   cdef char cdiag = 'U' if diag else 'N'
   cdef int lda = m if left_side else n
   cdef int ldb = m
-  strsm(&cside, &cuplo, &ctransa, &cdiag, &m, &n, alpha, a, &lda, x, &ldb)
+
+  for _ in range(num_b):
+    strsm(&cside, &cuplo, &ctransa, &cdiag, &m, &n, alpha, a, &lda, x, &ldb)
+    x += n*m
```

Comment:
You probably should cast `n` and `m` to 64-bit integers, otherwise you may see integer overflows if the matrix is over 2^32 elements. Same everywhere.

Reply:
I've cast the product `m*n` to a 64-bit integer. Bit confused about casting `m` and `n`.

m and n refer to the size of last two dimensions of the matrix, and the rest of the file seems to treat these parameters are 32-bit.

 I think they'd only need to be be size 2^64 is the matrix had over 2^4096 it its two final dimension?

Reply:
LAPACK itself seems to use 32-bit integers for the dimension sizes. We can't easily change that, although one might argue it should be using 64-bit integers.

LAPACK's choice limits the size of any individual dimension of the array to 2^32. However, it's quite possible (and indeed, users have previously hit the case) to have an array with dimensions > 2^16. The product of two such arrays has a total size > 2^32. So whenever we are taking the product of two array dimensions, we must be careful to do the multiplication on 64-bit integers.

It would be simpler if we could just use 64-bit integers everywhere, but then we'd have to cast back to 32-bit for passing the values to LAPACK.